### PR TITLE
Code folding

### LINKED ISSUE #4083: Code folding
Currently when I click on the blue vertical bar next to a cell only 3 dots remain. That way all the info about that cell is lost. I don't see an option to enable code folding (e.g. if the first line contains a comment, leave that comment there and also let the cell be executable).
Are folding vs. collapsing things that should be handled completely separate? If so, is there a plan to implement code folding? Especially for long cells with only python import statements that would be extremely useful in my opinion.

### PR DESCRIPTION
**Pull request by fcollonval**
Fixes #4083

**Screenshots**
codefolding

### CODE REVIEW

File: packages/codeeditor/src/editor.ts [hunk 1]
```diff
@@ -654,7 +659,8 @@ export namespace CodeEditor {
     tabSize: 4,
     insertSpaces: true,
     matchBrackets: true,
-    autoClosingBrackets: true
+    autoClosingBrackets: true,
+    codeFolding: true
```
Comment (ian-r-rose) [inline packages/codeeditor/src/editor.ts]: As with #5750, I think we may want to have this `false` by default (at least to start).

File: packages/fileeditor-extension/schema/plugin.json [hunk 2]
```diff
@@ -63,7 +66,8 @@
         "readOnly": false,
         "insertSpaces": true,
         "tabSize": 4,
-        "wordWrapColumn": 80
+        "wordWrapColumn": 80,
+        "codeFolding": true
```
Comment (ian-r-rose) [inline packages/fileeditor-extension/schema/plugin.json]: And here.

File: packages/notebook-extension/schema/tracker.json [hunk 3]
```diff
@@ -270,7 +273,8 @@
         "readOnly": false,
         "insertSpaces": true,
         "tabSize": 4,
-        "wordWrapColumn": 80
+        "wordWrapColumn": 80,
+        "codeFolding": false
       }
     },
     "markdownCellConfig": {
```
Comment (ian-r-rose) [inline packages/notebook-extension/schema/tracker.json]: As with #5750, not having this option in the markdown/raw cell settings makes it unconditionally on for those cells.

File: packages/codemirror/src/editor.ts [hunk 4]
```diff
@@ -1368,6 +1381,30 @@ namespace Private {
...
         break;
+      case 'lineNumbers':
+        let currentGutters = editor.getOption('gutters').slice() as Array<
+          string
+        >;
+        const lineNoIndex = currentGutters.indexOf('CodeMirror-linenumbers');
+        if (lineNoIndex < 0 && value) {
+          currentGutters.unshift('CodeMirror-linenumbers');
+        } else if (lineNoIndex === 0 && !value) {
+          currentGutters.shift();
```
Comment (ian-r-rose) [inline packages/codemirror/src/editor.ts]: Are there any other gutters that people might want to add? This logic could get difficult to read if there are more than two possible gutters. I am not sure if there is a better way, however, so happy to leave it as is...

File: packages/codemirror/src/editor.ts [hunk 5]
```diff
@@ -1316,6 +1329,23 @@ namespace Private {
...
+   */
+  function getActiveGutters(config: CodeMirrorEditor.IConfig): string[] {
+    // The order of the classes will be the gutters order
+    let classToSwitch = {
+      'CodeMirror-linenumbers': 'lineNumbers',
+      'CodeMirror-foldgutter': 'codeFolding'
+    };
+    return Object.keys(classToSwitch).filter(
+      // @ts-ignore
+      gutter => config[classToSwitch[gutter]]
```
Comment (fcollonval) [inline packages/codemirror/src/editor.ts]: Do you know how I can get ride of the `ts-ignore`? I don't know how to get a `ts-lint` compliant code here.

**Comment by fcollonval** [comment]
I went for the new proposal. What do you think?

**Comment by ian-r-rose** [comment]
@fcollonval pinging on this

**Comment by fcollonval** [comment]
@ian-r-rose here it is :wink:

**Pull request merged**